packages/contracts-da-bridge/contracts/DABridgeRouter.sol

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity 0.7.6;
+
+// ============ External Imports ============
+import {TypedMemView} from "@summa-tx/memview-sol/contracts/TypedMemView.sol";
+// ============ Internal Imports ============
+import {DABridgeMessage} from "./DABridgeMessage.sol";
+import {XAppConnectionClient} from "@nomad-xyz/contracts-router/contracts/XAppConnectionClient.sol";
+import {Router} from "@nomad-xyz/contracts-router/contracts/Router.sol";
+import {Version0} from "@nomad-xyz/contracts-core/contracts/Version0.sol";
+
+/**
+ * @title DABridgeRouter
+ */
+contract DABridgeRouter is Version0, Router {
+    // ============ Libraries ============
+
+    using TypedMemView for bytes;
+    using TypedMemView for bytes29;
+    using DABridgeMessage for bytes29;
+
+    // ============ Public Storage ============
+
+    mapping(uint64 => bytes32) public roots;
+
+    // ============ Upgrade Gap ============
+
+    // gap for upgrade safety
+    uint256[50] private __GAP;
+
+    // ============ Events ============
+
+    /**
+     * @notice emitted when a new data root is received
+     * @param originAndNonce Domain where the transfer originated and the
+     *        unique identifier for the message from origin to destination,
+     *        combined in a single field ((origin << 32) & nonce)
+     * @param blockNumber for data root
+     * @param root data root
+     */
+    event Receive(
+        uint64 indexed originAndNonce,
```

Comment:
what is the purpose of passing origin and nonce here? why not just origin? is nonce used by any off-chain applications in this case?

Reply:
ah i guess they are using it to link the `Receive` event definitively to the message that caused it, since this is the unique identifier for a message. you should be able to just observe the Process(messageHash) event that fires directly after Receive, but fair enough if you want it explicitly linked within this data